File: leetcode/go/1071.greatest-common-divisor-of-strings.go
func gcdOfStrings(str1 string, str2 string) string {
    gcd_len := gcd(len(str1), len(str2))
    common_substring := str1[:gcd_len]
    for i:=gcd_len; i < len(str1); i+=gcd_len {
        if str1[i:i+gcd_len] != common_substring {
            return ""
        }
    }
    for i:=0; i < len(str2); i+=gcd_len {
        if str2[i:i+gcd_len] != common_substring {
            return ""
        }
    }
    return common_substring
}

func gcd(a int, b int) int {
    if b == 0 {return a}
    return gcd(b, a % b)
}

File: leetcode/python/1071.greatest-common-divisor-of-strings.py
class Solution:
    def gcdOfStrings(self, str1: str, str2: str) -> str:
        
        def gcd(a: int, b: int) -> int:
            if b == 0:
                return a
            return gcd(b, a % b)
        
        gcdLen = gcd(len(str1), len(str2))

        commonString = str1[:gcdLen]

        for i in range(0, len(str1), gcdLen):
            if commonString != str1[i:i+gcdLen]:
                return ""
        
        for i in range(0, len(str2), gcdLen):
            if commonString != str2[i:i+gcdLen]:
                return ""

        return commonString

""" """

class Solution:
    def gcdOfStrings(self, str1: str, str2: str) -> str:
        
        if str1 + str2 != str2 + str1:
            return ""

        def gcd(a: int, b: int) -> int:
            if b == 0:
                return a
            return gcd(b, a % b)
        
        gcdLen = gcd(len(str1), len(str2))

        return str1[:gcdLen]

